charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,259 @@
+K8S_RESOURCES = {
+    "template_files": ["src/templates/auth_manifests.yaml.j2"],
+    "resource_types": {ClusterRole, ClusterRoleBinding},
+    "scope": "tensorboard",
+}
+CRD_RESOURCES = {
+    "template_files": ["src/templates/crds.yaml.j2"],
+    "resource_types": {CustomResourceDefinition},
+    "scope": "tensorboard",
+}
 
-    def __init__(self, msg, status_type=None):
-        super().__init__()
+TENSORBOARD_IMAGE = "tensorflow/tensorflow:2.1.0"
 
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
 
+class TensorboardController(CharmBase):
+    """Tensorboard Controller Charmed Operator."""
 
-class Operator(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
+        self.logger = logging.getLogger(__name__)
+
+        # retrieve configuration and base settings
+        self._namespace = self.model.name
+        self._lightkube_field_manager = "lightkube"
+        self._name = self.app.name
+        self._exec_command = "/manager"
+        self._container_name = "tensorboard-controller"
+        self._container = self.unit.get_container(self._container_name)
+
+        # setup context to be used for updating K8s resources
+        self._context = {
+            "app_name": self._name,
+            "namespace": self._namespace,
+            "service": self._name,
+        }
+        self._rbac_resource_handler = None
+        self._crd_resource_handler = None
+
         self.gateway = GatewayRequirer(self)
 
-        for event in [
-            self.on.install,
-            self.on.leader_elected,
-            self.on.upgrade_charm,
-            self.on.config_changed,
-            self.on["gateway-info"].relation_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        # setup events to be handled by main event handler
+        self.framework.observe(self.on.config_changed, self._on_event)
+        self.framework.observe(self.on.leader_elected, self._on_event)
+        self.framework.observe(self.on.tensorboard_controller_pebble_ready, self._on_event)
+        self.framework.observe(self.on["gateway-info"].relation_changed, self._on_event)
 
-    def main(self, event):
-        try:
-            self._check_leader()
+        # setup events to be handled by specific event handlers
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade)
+        self.framework.observe(self.on.remove, self._on_remove)
 
-            image_details = self._check_image_details()
-        except CheckFailed as check_failed:
-            self.model.unit.status = check_failed.status
-            return
+    @property
+    def container(self) -> Container:
+        """Return tensorboard-controller container object."""
+        return self._container
```

Comment:
Is this property necessary?  It doesn't do anything when it returns and there's no associated setter.

Reply:
We use this pattern everywhere, including other charms and chisme. I've given it some thought and I believe this is a Python convention for providing read-only properties:
* Define "private" (i.e. undescore prefixed) attribute
* Make it public (readable) by exposing it through a property
* Provide a setter if we also want it to be writable